### PR TITLE
Add HyperlinkButton Control

### LINKED ISSUE #7640: Add HyperlinkButton
**Is your feature request related to a problem? Please describe.**

Other XAML frameworks have a [HyperlinkButton](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.hyperlinkbutton) (not to be confused with an inline hyperlink text element inside of a text control). This control is useful to:

 1. Simplify the common use case of a hyperlink button
 2. Provide a standard look-and-feel for hyperlinks in the framework
 3. Provide a cross-platform way of navigating to Uri like webpages.
 4. Automatically manages a change in text color after a link is clicked the first time.

There are additional opportunities here as well. The `HyperlinkButton` could have a new `NavigateUri` property added (like UWP). If a NavigateUri is set, then the HyperlinkButton is pressed, that Uri would be opened in a platform-specific way. This would greatly simplify opening something like a webpage on each platform.

**Describe the solution you'd like**

Add a new HyperlinkButton control derived from button with a standardized style.

**Describe alternatives you've considered**

I've seen often times in the past developers are told to just create their own custom Button style and use that instead.

**Additional context**

### PR DESCRIPTION
**Pull request by rabbitism**
## What does the pull request do?
Add HyperlinkButton control

## How was the solution implemented (if it's not obvious)?
HyperlinkButton inherits from Button with three additional properties:
1. DirectProperty `NavigateUri`: The Uri to navigate to upon click. 
2. StyledProperty `IsVisited`: Indicate whether this Uri is already visited. 
3. StyledProperty `IsVisitedSetOnClick`: True means `IsVisited` automatically set to true upon click. False means `IsVisited` always take assigned value. 

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes #7640 

## Comments
Notice that I added unit test and tested locally on Windows and all passed. But I commented them out to avoid calling `Process.Start` in CI/CD.

### CODE REVIEW

File: src/Avalonia.Controls/HyperlinkButton.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+using Avalonia.Logging;
+using Avalonia.Styling;
+
+namespace Avalonia.Controls
+{
+    public class HyperlinkButton: Button, IStyleable
+    {
+        public const string pcVisited = ":visited";
+
+        Type IStyleable.StyleKey => typeof(HyperlinkButton);
```
Comment (maxkatz6) [inline src/Avalonia.Controls/HyperlinkButton.cs:17]: by default it returns `this.GetType()`, so I don't think it's necessary

File: tests/Avalonia.Controls.UnitTests/HyperlinkButtonTests.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    public void HyperlinkButton_IsVisitedSetOnClick_Enabled()
+    {
+        var hyperlinkButton = new TestHyperlinkButton()
+        {
+            IsVisitedSetOnClick = true, 
+            NavigateUri = new Uri("test://test"),
+        };
+        // hyperlinkButton.OnClick();
+        // Assert.True(hyperlinkButton.IsVisited);
+        // Assert.True(hyperlinkButton.Classes.Contains(HyperlinkButton.pcVisited));
```
Comment (robloo) [inline tests/Avalonia.Controls.UnitTests/HyperlinkButtonTests.cs:32]: Should all of this commented-out code stay here and below?

**Review by robloo — COMMENTED** [review]
Awesome work! Thanks for picking up this control!

**Comment by danwalmsley** [comment]
> I am not sure if we should include "Process.Open" in the default control. Even more I am not sure if it works crossplat. it should be configurable. Probably `Command` can by default has a command to run `Process.Open` and `CommandParameter` by default can have value of the `NavigateUri`.
> 
> Also, why is `IsVisited` mutable? It can be a readonly direct property. Upd, I see there is example of usage, when IsVisited set in xaml. Still not sure if it needs to be "styled", but fine by me to be mutable.


It wont work cross platform: here is a cross platform implementation, that will work!

you need to implement "shell execution" and "openbrowserasync".

these probably should end up in some kind of platform api for avalonia.
https://github.com/zkSNACKs/WalletWasabi/blob/v1.1.13.1/WalletWasabi/Helpers/IoHelpers.cs#L144-L162

**Comment by rabbitism** [comment]
Updated to fix the formatting issue. Will check further on cross platform implementation.

**Comment by rabbitism** [comment]
Is this a good reference?

https://github.com/dotnet/maui/tree/main/src/Essentials/src/Launcher

**Comment by maxkatz6** [comment]
@rabbitism yes. iOS and Android implementation can use same APIs as there, everything else can be reused from code @danwalmsley attached (as it can be reused across main desktop platforms).

**Comment by rabbitism** [comment]
So, before I implement anything, can we have a discussion of whether we should have a Launcher like feature in Avalonia? It might be similar to IClipboard. I think it will benefit not only this control.

**Comment by maxkatz6** [comment]
@rabbitism we can have API like this (similar to Essentials):
```
	public interface ILauncher
	{
		Task<bool> CanOpenAsync(Uri uri);
		Task<bool> OpenAsync(Uri uri);
		Task<bool> OpenAsync(IStorageFile file);
	}
```
Which can be accessible by `AvaloniaLocator.GetService<ILauncher>()` (until we have better API for the services)
Not sure if we need PresentationSourceBounds like MAUI has, not familiar with that.

**Comment by tomenscape** [comment]
Hyperlinks are not necessarily used to immediately open a URL in an external process.

In my WPF application we have various cases where clicking a hyperlink executes an event handler. Often the event handler does _eventually_ launch a web browser, but not before some pre-processing steps are executed (e.g. authentication token requests).  The URL is not known to the hyperlink.

There are other cases where a hyperlink just takes the user to a different area of the current application, and in this case the hyperlink again has no URL.

Even when clicking a hyperlink does immediately open a URL, we have an process launching interface of our own which does this, and would not want to use a Button which bypasses it.

If built-in process launching is to be added to Avalonia, I would recommend that the button raise a `RoutedEvent` that has a built-in class handler for `TopLevel` controls. This gives users the option of intercepting the event and handling it themselves. The event should be raised regardless of whether a URL is defined on the hyperlink.

**Comment by robloo** [comment]
@tomenscape This is fully configurable to do what you want. If you don't specify a NavigateUrl nothing will happen on click. You can also disable automatic link color change with IsVisitedSetOnClick=False.

**Comment by rabbitism** [comment]
Close for now as this is really outdated codebase, and I need to reevaluate the effort to implement ILauncher.

**Pull request closed without merging**